### ISSUE TITLE
error: cannot use context manager

### ISSUE THREAD

**Issue by epogrebnyak** [mini-kep/db#3]
```python 

def _find_by(session_factory, condition=None):
    # FIXME: why not working with context manager?
    # DetachedInstanceError: Instance <Datapoint at 0x9074ef0> 
    # is not bound to a Session; attribute refresh operation cannot proceed
    with Session(session_factory) as session:
        query = session.query(Datapoint)
        if condition is not None:
            return query.filter_by(**condition).all()
        else:
            return query.all()    
```

**Comment by epogrebnyak** [comment]
call to failing function:

```
_find_by(session_factory, condition=dict(date="2014-03-31", freq='q', name="CPI_rog"))
```

**Comment by varnie** [comment]
In the current implementation of class Session (context) actual commit occurs upon `__exit__`'ing only.
but `def insert_one` `returns True` BEFORE the commit has been performed. It doesn't work this way.

Here's slightly changed version which handles this error properly:
```
# session / row operations
def insert_one(session_factory, datapoint):
    with Session(session_factory) as session:
        #session.add(datapoint)
        #return True
        try:
            session.add(datapoint)
            session.commit()
            return True
        except:
            session.rollback()
            return False
```

It is not 100% accurate with respect to current implementation of `class Session`, but I hope it demonstrates the source of the bug.

In other words, in order to catch Integrity-related problems (and all the others, if there're any) when we perform insert, we have to perform commit after adding, not later and try/catch the exceptions.
The current implementations of `def insert_one` silently returns True ALWAYS and doesn't bother to handle exceptions.

**Comment by epogrebnyak** [comment]
I agree on exception handling, but do not really like bare ```except```, that catches everything, generally not great practice. 

I tried picking some excpetions from sqlalchemy, but pytest did not quite catch them.

**Comment by epogrebnyak** [comment]
<https://github.com/mini-kep/db/blob/master/clientdb.py> is where it all happens

**Comment by epogrebnyak** [comment]
A submission from @Rub4ek: 
```
@contextmanager
def transactional_scope(session_factory):
    """Provide a transactional scope around a series of operations."""
    session = session_factory()
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()
```

**Comment by epogrebnyak** [comment]
I close this issue for now.
